### PR TITLE
Navmap rework

### PR DESCRIPTION
**Pull request by chromiumboy**
## About the PR

This PR adds additional functionality and optimizations to nav maps (see below for the full list). 

## Why / Balance

On the player-facing side, the nav map now displays finer structural details, like thin windows and doors, and should exhibit better performance 

## Technical details

Changes made
- The number of NavMapComponent dirty calls in NavMapSystem has been reduced. When a nav map chunk is updated, only the affected chunk is sent to the client
- Nav map chunks use ushorts instead of ints
- Airlocks data is transmitted as nav map chunks instead of entity lists
- Floor tile chunks are now also sent to the client - this will be needed for an upcoming PR (nav map regions)
- Nav map chunks have cartesian directional information embedded, to allow thin objects to be drawn on the nav map
- Colinear nav map lines are now combined to increase draw speed
- The code for rendering tracked entities has been cleaned up
- Airlocks and floor tile rendering data is only updated once a second, like walls, to increase draw speed
- Floor tiles vertices have been rounded to eliminate unsightly seams

## Media

Screenshot from the updated nav map, showing how thin windows and doors are depicted

navmap_update

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

The NavMapLine class was removed, as it was no longer needed - it can be replaced with a tuple pair of Vector2, i.e., '(Vector2, Vector2)', if required

**Changelog**

### CODE REVIEW

File: Content.Client/Pinpointer/NavMapSystem.cs [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
+public sealed partial class NavMapSystem : SharedNavMapSystem
 {
     public override void Initialize()
     {
         base.Initialize();
+
         SubscribeLocalEvent<NavMapComponent, ComponentHandleState>(OnHandleState);
+
+        SubscribeNetworkEvent<NavMapChunkChangedEvent>(OnChunkChanged);
+        SubscribeNetworkEvent<NavMapBeaconChangedEvent>(OnBeaconChanged);
```
[the remark this answers is not in the document]
Comment (chromiumboy) [reply Content.Client/Pinpointer/NavMapSystem.cs:16]: I'll look into it, but I haven't used delta states before and there aren't many examples to learn from

[the remark this answers is not in the document]
Comment (chromiumboy) [reply Content.Client/Pinpointer/NavMapSystem.cs]: OK, I did the thing, and it seems to be working. But I'd go through the delta state changes with a fine tooth comb, as its kinda fiddly...

File: Content.Shared/Pinpointer/SharedNavMapSystem.cs [hunk 2]
```diff
@@ -31,19 +56,235 @@ public static Vector2i GetTile(int flag)
...
+        return chunk;
+    }
+
+    public ushort GetCombinedEdgesForChunk(Dictionary<AtmosDirection, ushort> tile)
+    {
+        ushort combined = 0;
+
+        foreach (var (_, value) in tile)
+            combined |= value;
+
```
Comment (metalgearsloth) [inline Content.Shared/Pinpointer/SharedNavMapSystem.cs:88]: Use .Keys and .Values for these as it's faster than this for whatever reason.

File: Content.Shared/Pinpointer/SharedNavMapSystem.cs [hunk 3]
```diff
@@ -31,19 +56,235 @@ public static Vector2i GetTile(int flag)
...
 
-        public List<NavMapAirlock> Airlocks = new();
+        string? name = component.Text;
+
+        if (string.IsNullOrEmpty(name))
+            name = MetaData(uid).EntityName;
+
+        beaconData = new NavMapBeacon(GetNetEntity(uid), component.Color, name, xform.LocalPosition)
+        {
+            LastUpdate = _gameTiming.CurTick
```
Comment (metalgearsloth) [inline Content.Shared/Pinpointer/SharedNavMapSystem.cs:132]: You should be able to use the metadatacomponent across both the entname and getnetentity.

**Comment by chromiumboy** [comment]
This PR request has been approved to be merged by Sloth

**Pull request merged**